src/main/java/javaslang/collection/TreeSet.java

```diff
@@ -92,13 +92,14 @@
 
     public static <T extends Comparable<? super T>> TreeSet<T> ofAll(java.lang.Iterable<? extends T> values) {
         Objects.requireNonNull(values, "values is null");
-        return new TreeSet<>(RedBlackTree.ofAll(values));
+        return values.iterator().hasNext() ? new TreeSet<>(RedBlackTree.ofAll(values)) : empty();
```

Comment:
You are right. hasNext() has also to be called by next() as of the API contract in Java docs. That means that it can be called twice or more without changing the public visible state. Thx for clarification.